### DISCUSSION TITLE
Counting image frames: is this a reliable algorithm?

### DISCUSSION THREAD

**Discussion by rhymes** [libvips/ruby-vips#282] [Q&A]
I would like to count frames of loaded images, regardless if they are animated or not. 

I came up with this (mostly borrowed by @jcupitt in https://github.com/libvips/ruby-vips/discussions/274#discussioncomment-325027):

```ruby
require "vips"

path = ARGV[0]
img = Vips::Image.new_from_file(path, access: :sequential)

frame_count = if img.get_typeof("n-pages") != 0
                img.get("n-pages")
              else
                page_height = if img.get_typeof("page-height") != 0
                                img.get("page-height")
                              else
                                img.height
                              end

                img.height / page_height
              end

puts "#{path}: #{frame_count}"

# $ vips_frame_count.rb image.gif 
# 19

# $ vips_frame_count.rb image.png
# 1
```

Is this reliable? Thank you!

**Comment by jcupitt** [comment]
Hello @rhymes,

`n-pages` always gives you the number of pages in the document, not the number of pages in the image. For example:

```
irb(main):001:0> require "vips"
=> true
irb(main):002:0> x = Vips::Image.new_from_file "nipguide.pdf"
irb(main):003:0> x.height
=> 842
irb(main):004:0> x.get("n-pages")
=> 58
```

So even though I only loaded the first page, I still see `n-pages` of 58.

It's best to divide height by page height, with page height defaulting to height if not defined. Eg.:

```ruby
def get_defaulted(image, name, default)
    if image.get_typeof(name) != 0
        image.get(name)
    else
        default
    end
end

def get_page_height(image)
    get_defaulted(image, "page-height", image.height)
end

def get_n_loaded_pages(image)
    image.height / get_page_height(image)
end
```

This should probably be added to ruby-vips (this logic is in other bindings, like the C API).

**Reply by rhymes** [reply]
Thank you for the correction! :)

**Reply by rhymes** [reply]
**update**:

I tested this with a few animated GIFs and it always returns `1`, that's because their `page-height` attribute does not exist.

Taking this one:

[gif's of muppet working](https://user-images.githubusercontent.com/146201/109296239-48cba100-7830-11eb-9ffe-38d684413770.gif)

Minimagick's `identify` on the command line finds 19 frames, by opening it in Ruby it still counts `19` frames:

```ruby
> MiniMagick::Image.new("image.gif").frames.count
=> 19
```

but if I use Vips I get:

```ruby
> image = Vips::Image.new_from_file("image.gif")
=> #<Image 500x280 uchar, 4 bands, srgb>
> image.get("page-height")
Vips::Error: vips_image_get: field "page-height" not found
```

thus the above algorithm defaults to dividing the height by itself, hence `1`.

Maybe my algorithm should be: "if `page-height / height` is 1, then check `n-pages`" ?

**Comment by jcupitt** [comment]
It depends what you want.

image height / page height gives you the number of loaded pages, n-pages gives you the number of pages in the underlying file. Eg.:

```
$ irb
irb(main):001:0> require "vips"
=> true
irb(main):002:1* def get_defaulted(image, name, default)
irb(main):003:2*   if image.get_typeof(name) != 0
irb(main):004:2*     image.get(name)
irb(main):005:2*   else
irb(main):006:2*     default
irb(main):007:1*   end
irb(main):008:0> end
=> :get_defaulted
irb(main):009:0> 
irb(main):010:1* def get_page_height(image)
irb(main):011:1*   get_defaulted(image, "page-height", image.height)
irb(main):012:0> end
=> :get_page_height
irb(main):013:0> 
irb(main):014:1* def get_n_loaded_pages(image)
irb(main):015:1*   image.height / get_page_height(image)
irb(main):016:0> end
=> :get_n_loaded_pages
irb(main):017:0> x = Vips::Image.new_from_file "3198.gif"
irb(main):018:0> x.get("n-pages")
=> 139
irb(main):019:0> get_n_loaded_pages(x)
=> 1
```

So I've only loaded 1 page, though the underlying file has 139 pages. I can load more pages, eg.:

```
irb(main):020:0> x = Vips::Image.new_from_file "3198.gif", page: 12, n: 14
irb(main):021:0> x.get("n-pages")
=> 139
irb(main):023:0> get_n_loaded_pages(x)
=> 14
```

Now I've loaded 14 pages, starting from page 12. You can use `n: -1` to load all pages.

**Reply by rhymes** [reply]
Great explanation. Thank you again @jcupitt!

**Comment by Nakilon** [comment]
Just recently there was a pull request in another gem -- https://github.com/sdsykes/fastimage/pull/119
They wanted to know "if the image is animated" and they decided to do it by checking if there are multiple frames. I told them that it does not mean it's animated and proposed to just return the number of frames. But they did it anyway: https://github.com/sdsykes/fastimage/blob/a4407988544b8b83e10b3466747b324cdc8c6f40/lib/fastimage.rb#L604-L605
One reply was that "we want library to be fast and not fetch a lot" but it's Ruby -- you can return the Enumerator of frames and take as many of them as you need...

**Reply by jcupitt** [reply]
Unfortunately, GIF does not record the number of frames in the header. You need to parse the whole file and count the number of `IMAGE_DESC` objects :-( It's a really awful format.

The libvips GIF loader works in two phases: on open() it tries to scan the whole file as quickly as possible, counting objects and checking for transparency and for colour (these are not recorded in the GIF header either, stupidly). In the second load() phase it decompresses the file a frame at a time as you read it. 

It's not great, but I can't think of a better solution, sadly.
